### PR TITLE
ICU-22407 Implement Java Temporal Calendar API

### PR DESCRIPTION
**Pull request by FrankYFTang**
##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22407
- [X] Required: The PR title must be prefixed with a JIRA Issue number. 
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [X] Issue accepted (done by Technical Committee after discussion)
- [ ] Tests included, if applicable
- [X] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

**Comment by FrankYFTang** [comment]
Since [richgillam](https://github.com/richgillam) reviewed https://github.com/unicode-org/icu/pull/2274 it probably would be easier for him to be the reviewer since it is a 1:1 porting from the C++ implementation

**Review by richgillam — DISMISSED** [review]
I probably should have read this more closely, but I'm running out of brain cells...

This looks good, and it looks like what I remember from the C side.

File: icu4j/main/classes/core/src/com/ibm/icu/util/CECalendar.java [hunk 1]
```diff
@@ -273,4 +275,49 @@ public static void jdToCE(int julianDay, int jdEpochOffset, int[] fields) {
...
+     * The value is a string identifier that starts with the literal grapheme
+     * "M" followed by two graphemes representing the zero-padded month number
+     * of the current month in a normal (non-leap) year. For the short thirteen
+     * month in each year in the CECalendar, the value is "M13".
+     *
+     * @return       One of 13 possible strings in {"M01".. "M12", "M13"}.
+     * @draft ICU 74
+     */
+    public String getTemporalMonthCode() {
+        if (get(MONTH) == 12) return "M13";
```
[the remark this answers is not in the document]
Comment (FrankYFTang) [reply icu4j/main/classes/core/src/com/ibm/icu/util/CECalendar.java:294]: good catch, the gTemporalMonthCodes should not be here.

File: icu4c/source/i18n/hebrwcal.cpp [hunk 2]
```diff
@@ -777,7 +777,7 @@ int32_t HebrewCalendar::internalGetMonth() const {
         HebrewCalendar *nonConstThis = (HebrewCalendar*)this; // cast away const
 
         int32_t year = nonConstThis->handleGetExtendedYear();
-        return ordinalMonth + ((isLeapYear(year) && (ordinalMonth > ADAR_1)) ? 1: 0);
+        return ordinalMonth + (((!isLeapYear(year)) && (ordinalMonth > ADAR_1)) ? 1: 0);
```
[the remark this answers is not in the document]
Comment (FrankYFTang) [reply icu4c/source/i18n/hebrwcal.cpp]: just make it more explicit that the ! is not applying to A&&B but only to A

[the remark this answers is not in the document]
Comment (FrankYFTang) [reply icu4c/source/i18n/hebrwcal.cpp]: and the cpp part is already merged in other PR and removed from this PR

**Comment by FrankYFTang** [comment]
PTAL

**Review by richgillam — APPROVED** [review]
Looks good.  Thanks for the fixes!

**Comment by FrankYFTang** [comment]
Have post-merge breakage on the following three tests

Testcase: TestConsistencyChinese took 10.636 sec
	FAILED
Round trip conversion produces different time from Mon Feb 17 18:06:15 PST 1890 to  Wed Feb 27 18:06:15 PST 1889 delta: -30672000000 Gregorian(e=1 1890/2/18) 
junit.framework.AssertionFailedError: Round trip conversion produces different time from Mon Feb 17 18:06:15 PST 1890 to  Wed Feb 27 18:06:15 PST 1889 delta: -30672000000 Gregorian(e=1 1890/2/18) 
	at com.ibm.icu.dev.test.AbstractTestLog.errln(AbstractTestLog.java:50)
	at com.ibm.icu.dev.test.calendar.IBMCalendarTest.checkConsistency(IBMCalendarTest.java:2162)
	at com.ibm.icu.dev.test.calendar.IBMCalendarTest.TestConsistencyChinese(IBMCalendarTest.java:2077)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)

Testcase: TestConsistencyDangi took 9.44 sec
	FAILED
Round trip conversion produces different time from Mon Feb 17 18:06:29 PST 1890 to  Wed Feb 27 18:06:29 PST 1889 delta: -30672000000 Gregorian(e=1 1890/2/18) 
junit.framework.AssertionFailedError: Round trip conversion produces different time from Mon Feb 17 18:06:29 PST 1890 to  Wed Feb 27 18:06:29 PST 1889 delta: -30672000000 Gregorian(e=1 1890/2/18) 
	at com.ibm.icu.dev.test.AbstractTestLog.errln(AbstractTestLog.java:50)
	at com.ibm.icu.dev.test.calendar.IBMCalendarTest.checkConsistency(IBMCalendarTest.java:2162)
	at com.ibm.icu.dev.test.calendar.IBMCalendarTest.TestConsistencyDangi(IBMCalendarTest.java:2082)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)


Testcase: TestConsistencyIndian took 1.273 sec
	FAILED
Round trip conversion produces different time from Fri Apr 26 18:06:44 PST 4713 to  Fri May 03 18:06:44 PST 4713 delta: 604800000 Gregorian(e=0 4713/4/27) 
junit.framework.AssertionFailedError: Round trip conversion produces different time from Fri Apr 26 18:06:44 PST 4713 to  Fri May 03 18:06:44 PST 4713 delta: 604800000 Gregorian(e=0 4713/4/27) 
	at com.ibm.icu.dev.test.AbstractTestLog.errln(AbstractTestLog.java:50)
	at com.ibm.icu.dev.test.calendar.IBMCalendarTest.checkConsistency(IBMCalendarTest.java:2162)
	at com.ibm.icu.dev.test.calendar.IBMCalendarTest.TestConsistencyIndian(IBMCalendarTest.java:2027)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)

**Comment by FrankYFTang** [comment]
The post merge breakage are the same as the C++ one in https://unicode-org.atlassian.net/browse/ICU-22252
Will work oon a new PR for Java based on https://github.com/unicode-org/icu/pull/2298 to address this soon

**Pull request merged**